src/raft/raft.rs

```diff
@@ -1764,26 +1822,50 @@ impl<T: Storage> Raft<T> {
     // promotable indicates whether state machine can be promoted to leader,
     // which is true when its own id is in progress list.
     pub fn promotable(&self) -> bool {
-        self.get_prs().contains_key(&self.id)
+        self.get_prs().voters().contains_key(&self.id)
     }
 
-    pub fn add_node(&mut self, id: u64) {
+    fn add_voter_or_learner(&mut self, id: u64, is_learner: bool) {
         self.pending_conf = false;
-        if self.get_prs().contains_key(&id) {
-            // Ignore any redundant addNode calls (which can happen because the
-            // initial bootstrapping entries are applied twice).
+        if self.get_prs().voters().contains_key(&id) {
+            if is_learner {
+                info!(
+                    "{} ignored add learner: do not support changing {} from voter to learner",
+                    self.tag,
+                    id
+                );
+            }
+            return;
+        }
+
+        if !is_learner && self.mut_prs().promote_learner(id) {
+            // Promote learner to voter success.
             return;
         }
+
+        // New progress.
         let last_index = self.raft_log.last_index();
-        self.set_progress(id, 0, last_index + 1);
+        self.set_progress(id, 0, last_index + 1, is_learner);
```

Comment:
Should it be ignored if is_learner is true, and `self.get_prs().learners().contains_key(&id)`?